README.md
# http2-proxy

A simple http/2 & http/1.1 to http/1.1 spec compliant proxy helper for Node.

## Features

- Proxies HTTP 2, HTTP 1 and WebSocket.
- Simple and high performance.
- [Hop by hop header handling](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers).
- [Connection header handling](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection).
- [Via header handling](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Via).
- [Forward header handling](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forward).

## Installation

```bash
$ npm install http2-proxy
```

## Notes

`http2-proxy` requires at least node **v10.0.0**.

Fully async/await compatible and all callback based usage is optional and discouraged.

During 503 it is safe to assume that the request never made it to the upstream server. This makes it safe to retry non idempotent methods.

Use a final and/or error handler since errored responses won't be cleaned up automatically. This makes it possible to perform retries.

```js
const finalhandler = require('finalhandler')

const defaultWebHandler = (err, req, res) => {
  if (err) {
    console.error('proxy error', err)
    finalhandler(req, res)(err)
  }
}

const defaultWSHandler = (err, req, socket, head) => {
  if (err) {
    console.error('proxy error', err)
    socket.destroy()
  }
}
```

## HTTP/1 API

You must pass `allowHTTP1: true` to the `http2.createServer` or `http2.createSecureServer` factory methods.

```js
import http2 from 'http2'
import proxy from 'http2-proxy'

const server = http2.createServer({ allowHTTP1: true })
server.listen(8000)
```

You can also use `http-proxy2` with the old `http` && `https` API's.

```js
import http from 'http'

const server = http.createServer()
server.listen(8000)
```

## API

### Proxy HTTP/2, HTTP/1 and WebSocket

```js
server.on('request', (req, res) => {
  proxy.web(req, res, {
    hostname: 'localhost'
    port: 9000
  }, defaultWebHandler)
})
server.on('upgrade', (req, socket, head) => {
  proxy.ws(req, socket, head, {
    hostname: 'localhost'
    port: 9000
  }, defaultWsHandler)
})
```

### Use [Connect](https://www.npmjs.com/package/connect) & [Helmet](https://www.npmjs.com/package/helmet)

```js
const app = connect()
app.use(helmet())
app.use((req, res, next) => proxy
  .web(req, res, {
    hostname: 'localhost'
    port: 9000
  }, err => {
    if (err) {
      next(err)
    }
  })
)
server.on('request', app)
```

### Add x-forwarded Headers

```js
server.on('request', (req, res) => {
  proxy.web(req, res, {
    hostname: 'localhost'
    port: 9000,
    onReq: (req, { headers }) => {
      headers['x-forwarded-for'] = req.socket.remoteAddress
      headers['x-forwarded-proto'] = req.socket.encrypted ? 'https' : 'http'
      headers['x-forwarded-host'] = req.headers['host']
    }
  }, defaultWebHandler)
})
```

### Follow Redirects

```js
const http = require('follow-redirects').http

server.on('request', (req, res) => {
  proxy.web(req, res, {
    hostname: 'localhost'
    port: 9000,
    onReq: (req, options) => http.request(options)
  }, defaultWebHandler)
})
```

### Add Response Header

```js
server.on('request', (req, res) => {
  proxy.web(req, res, {
    hostname: 'localhost'
    port: 9000,
    onReq: (req, options) => http.request(options),
    onRes: (req, res, proxyRes) => {
      res.setHeader('x-powered-by', 'http2-proxy')
      res.writeHead(proxyRes.statusCode, proxyRes.headers)
      proxyRes.pipe(res)
    }
  }, defaultWebHandler)
})
```

### Try Multiple Upstream Servers (Advanced)

```js
const http = require('http')
const proxy = require('http2-proxy')
const createError = require('http-errors')

server.on('request', async (req, res) => {
  try {
    res.statusCode = null
    for await (const { port, timeout, hostname } of upstream) {
      if (req.aborted || finished) {
        return
      }

      let error = null
      let finished = false
      let bytesWritten = 0
      try {
        return await proxy.web(req, res, {
          port,
          timeout,
          hostname,
          onRes: async (req, res, proxyRes) => {
            if (proxyRes.statusCode >= 500) {
              throw createError(proxyRes.statusCode, proxyRes.message)
            }

            function setHeaders () {
              if (!bytesWritten) {
                res.statusCode = proxyRes.statusCode
                for (const [ key, value ] of Object.entries(headers)) {
                  res.setHeader(key, value)
                }
              }
            }

            // NOTE: At some point this will be possible
            // proxyRes.pipe(res)

            proxyRes
              .on('data', buf => {
                setHeaders()
                bytesWritten += buf.length
                if (!res.write(buf)) {
                  proxyRes.pause()
                }
              })
              .on('end', () => {
                setHeaders()
                // WORKAROUND: https://github.com/nodejs/node/pull/27984
                if (!proxyRes.aborted) {
                  res.end()
                  // WORKAROUND: https://github.com/nodejs/node/pull/24347
                  finished = true
                }
              })
              .on('close', () => {
                res.off('drain', onDrain)
              }))

            res.on('drain', onDrain)

            function onDrain () {
              proxyRes.resume()
            }
          }
        })
      } catch (err) {
        if (!err.statusCode) {
          throw err
        }

        error = err

        if (err.statusCode === 503) {
          continue
        }

        if (req.method === 'HEAD' || req.method === 'GET') {
          if (!bytesWritten) {
            continue
          }

          // TODO: Retry range request
        }

        throw err
      }
    }

    throw error || new createError.ServiceUnavailable()
  } catch (err) {
    defaultWebHandler(err)
  }
}
```

### `[async] web (req, res, options[, callback])`

- `req`: [`http.IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage) or [`http2.Http2ServerRequest`](https://nodejs.org/api/http2.html#http2_class_http2_http2serverrequest).
- `res`: [`http.ServerResponse`](https://nodejs.org/api/http.html#http_class_http_serverresponse) or [`http2.Http2ServerResponse`](https://nodejs.org/api/http2.html#http2_class_http2_http2serverresponse).
- `options`: See [Options](#options)
- `callback(err, req, res)`: Called on completion or error.

See [`request`](https://nodejs.org/api/http.html#http_event_request)

### `[async] ws (req, socket, head, options[, callback])`

- `req`: [`http.IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage).
- `socket`: [`net.Socket`](https://nodejs.org/api/net.html#net_class_net_socket).
- `head`: [`Buffer`](https://nodejs.org/api/buffer.html#buffer_class_buffer).
- `options`: See [Options](#options).
- `callback(err, req, socket, head)`: Called on completion or error.

See [`upgrade`](https://nodejs.org/api/http.html#http_event_upgrade)

### `options`

- `hostname`: Proxy [`http.request(options)`](https://nodejs.org/api/http.html#http_http_request_options_callback) target hostname.
- `port`: Proxy [`http.request(options)`](https://nodejs.org/api/http.html#http_http_request_options_callback) target port.
- `protocol`: Agent protocol (`'http'` or `'https'`). Defaults to `'http'`.
- `path`: Target pathname. Defaults to `req.originalUrl || req.url`.
- `proxyTimeout`: Proxy [`http.request(options)`](https://nodejs.org/api/http.html#http_http_request_options_callback) timeout.
- `proxyName`: Proxy name used for **Via** header.
- `timeout`: [`http.IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage) or [`http2.Http2ServerRequest`](https://nodejs.org/api/http2.html#http2_class_http2_http2serverrequest) timeout.
- `[async] onReq(req, options[, callback])`: Called before proxy request. If returning a truthy value it will be used as the request.
  - `req`: [`http.IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage) or [`http2.Http2ServerRequest`](https://nodejs.org/api/http2.html#http2_class_http2_http2serverrequest)
  - `options`: Options passed to [`http.request(options)`](https://nodejs.org/api/http.html#http_http_request_options_callback).
  - `callback(err)`: Called on completion or error.
- `[async] onRes(req, resOrSocket, proxyRes[, callback])`: Called on proxy response. Writing of response must be done inside this method if provided.
  - `req`: [`http.IncomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage) or [`http2.Http2ServerRequest`](https://nodejs.org/api/http2.html#http2_class_http2_http2serverrequest).
  - `resOrSocket`: For `web` [`http.ServerResponse`](https://nodejs.org/api/http.html#http_class_http_serverresponse) or [`http2.Http2ServerResponse`](https://nodejs.org/api/http2.html#http2_class_http2_http2serverresponse) and for `ws` [`net.Socket`](https://nodejs.org/api/net.html#net_class_net_socket).
  - `proxyRes`: [`http.ServerResponse`](https://nodejs.org/api/http.html#http_class_http_serverresponse).
  - `callback(err)`: Called on completion or error.

## Node

These are some existing issues in NodeJS to keep in mind when writing proxy code.

- https://github.com/nodejs/node/issues/27981
- https://github.com/nodejs/node/issues/28001
- https://github.com/nodejs/node/issues/27880
- https://github.com/nodejs/node/issues/24743
- https://github.com/nodejs/node/issues/24742

And some pending PR's:

- https://github.com/nodejs/node/pull/28004
- https://github.com/nodejs/node/pull/27984
- https://github.com/nodejs/node/pull/24347

Some of these are further referenced in the examples.

## License

  [MIT](LICENSE)


CHANGELOG.md
# 5.0.0

* Internal rewrite.
* Pass hop by hop processed headers to onReq callback.
* 503 error is guaranteed to not have written anything to res. This is useful for proxies which want to be able to retry against other upstream server.
* onRes and onReq can now return a promise (callback is optional).


index.js
const net = require('net')
const http = require('http')
const https = require('https')

const CONNECTION = 'connection'
const HOST = 'host'
const KEEP_ALIVE = 'keep-alive'
const PROXY_AUTHORIZATION = 'proxy-authorization'
const PROXY_AUTHENTICATE = 'proxy-authenticate'
const PROXY_CONNECTION = 'proxy-connection'
const TE = 'te'
const FORWARDED = 'forwarded'
const TRAILER = 'trailer'
const TRANSFER_ENCODING = 'transfer-encoding'
const UPGRADE = 'upgrade'
const VIA = 'via'
const AUTHORITY = ':authority'
const HTTP2_SETTINGS = 'http2-settings'

const kReq = Symbol('req')
const kRes = Symbol('res')
const kProxyCallback = Symbol('callback')
const kProxyReq = Symbol('proxyReq')
const kProxyRes = Symbol('proxyRes')
const kProxySocket = Symbol('proxySocket')
const kConnected = Symbol('connected')
const kOnRes = Symbol('onRes')

module.exports = proxy

proxy.ws = function ws (req, socket, head, options, callback) {
  const promise = compat({ req, socket, head }, options)
  if (!callback) {
    return promise
  }
  // Legacy compat...
  promise
    .then(() => callback(null, req, socket, head))
    .catch(err => callback(err, req, socket, head))
}
proxy.web = function web (req, res, options, callback) {
  const promise = compat({ req, res }, options)
  if (!callback) {
    return promise
  }
  // Legacy compat...
  promise
    .then(() => callback(null, req, res))
    .catch(err => callback(err, req, res))
}

async function compat (ctx, options) {
  const { req, res } = ctx

  const {
    hostname,
    port,
    protocol,
    timeout,
    proxyTimeout,
    proxyName,
    onReq,
    onRes
  } = options

  // Legacy compat...
  if (timeout != null) {
    req.setTimeout(timeout)
  }

  await proxy(
    { ...ctx, proxyName },
    async ureq => {
      ureq.hostname = hostname
      ureq.port = port
      ureq.timeout = proxyTimeout

      if (onReq) {
        if (onReq.length <= 2) {
          return onReq(req, options)
        } else {
          // Legacy compat...
          return new Promise((resolve, reject) => {
            const promiseOrReq = onReq(req, options, (err, val) => err ? reject(err) : resolve(val))
            if (promiseOrReq) {
              if (promiseOrReq.then) {
                promiseOrReq.then(resolve).catch(reject)
              } else if (promiseOrReq.abort) {
                resolve(promiseOrReq)
              } else {
                throw new Error('onReq must return a promise or a request object')
              }
            }
          })
        }
      } else {
        let agent
        if (protocol == null || /^(http|ws):?$/.test(protocol)) {
          agent = http
        } else if (/^(http|ws)s:?$/.test(protocol)) {
          agent = https
        } else {
          throw new HttpError(`invalid protocol`, null, 500)
        }
        return agent.request(options)
      }
    },
    async (proxyRes, headers) => {
      proxyRes.headers = headers
      if (onRes.length <= 3) {
        return onRes(req, res, proxyRes)
      } else {
        // Legacy compat...
        return new Promise((resolve, reject) => {
          const promise = onRes(req, res, proxyRes, (err, val) => err ? reject(err) : resolve(val))
          if (promise && promise.then) {
            promise.then(resolve).catch(reject)
          }
        })
      }
    }
  )
}

async function proxy ({ req, socket, res = socket, head, proxyName }, onReq, onRes) {
  let callback
  let promise = new Promise((resolve, reject) => {
    callback = err => err ? reject(err) : resolve()
  })

  const headers = getRequestHeaders(req, proxyName)

  if (head !== undefined) {
    if (req.method !== 'GET') {
      throw new HttpError('only GET request allowed', null, 405)
    }

    if (req.headers[UPGRADE] !== 'websocket') {
      throw new HttpError('missing upgrade header', null, 400)
    }

    if (head && head.length) {
      res.unshift(head)
    }

    setupSocket(res)

    headers[CONNECTION] = 'upgrade'
    headers[UPGRADE] = 'websocket'
  }

  const proxyReq = await onReq({
    method: req.method,
    path: req.originalUrl || req.url,
    headers
  })

  req[kRes] = res

  res[kReq] = req
  res[kRes] = res
  res[kProxySocket] = null
  res[kProxyReq] = proxyReq
  res[kProxyRes] = null
  res[kProxyCallback] = callback

  proxyReq[kReq] = req
  proxyReq[kRes] = res
  proxyReq[kConnected] = false
  proxyReq[kOnRes] = onRes

  res
    .on('close', onComplete)
    .on('finish', onComplete)
    .on('error', onComplete)

  req
    .on('close', onComplete)
    .on('aborted', onComplete)
    .on('error', onComplete)

  proxyReq
    .on('error', onProxyReqError)
    .on('timeout', onProxyReqTimeout)
    .on('response', onProxyReqResponse)
    .on('upgrade', onProxyReqUpgrade)

  deferToConnect.call(proxyReq, onProxyConnect)

  return promise
}

function deferToConnect (cb) {
  this.once('socket', function (socket) {
    if (!socket.connecting) {
      cb.call(this)
    } else {
      socket.once('connect', cb.bind(this))
    }
  })
}

function onComplete (err) {
  const res = this[kRes]
  const req = res[kReq]

  if (!res[kProxyCallback]) {
    return
  }

  const proxySocket = res[kProxySocket]
  const proxyReq = res[kProxyReq]
  const proxyRes = res[kProxyRes]
  const callback = res[kProxyCallback]

  res[kProxySocket] = null
  res[kProxyReq] = null
  res[kProxyRes] = null
  res[kProxyCallback] = null

  res
    .off('close', onComplete)
    .off('finish', onComplete)
    .off('error', onComplete)

  req
    .off('close', onComplete)
    .off('aborted', onComplete)
    .off('error', onComplete)

  if (proxySocket) {
    proxySocket.destroy()
  }

  if (proxyRes) {
    proxyRes.destroy()
  }

  if (proxyReq) {
    if (proxyReq.abort) {
      proxyReq.abort()
    } else if (proxyReq.destroy) {
      proxyReq.destroy()
    }
  }

  callback(err)
}

function onProxyConnect () {
  this[kConnected] = true
  this[kReq].pipe(this)
}

function onProxyReqError (err) {
  err.statusCode = this[kConnected] ? 502 : 503
  onComplete.call(this, err)
}

function onProxyReqTimeout () {
  onComplete.call(this, new HttpError('proxy timeout', 'ETIMEDOUT', 504))
}

async function onProxyReqResponse (proxyRes) {
  const res = this[kRes]

  res[kProxyRes] = proxyRes
  proxyRes[kRes] = res

  const headers = setupHeaders(proxyRes.headers)

  proxyRes
    .on('aborted', onProxyResAborted)
    .on('error', onProxyResError)

  if (this[kOnRes]) {
    try {
      await this[kOnRes](proxyRes, headers)
    } catch (err) {
      onComplete.call(this, err)
    }
  } else if (!res.writeHead) {
    if (!proxyRes.upgrade) {
      res.write(createHttpHeader(`HTTP/${proxyRes.httpVersion} ${proxyRes.statusCode} ${proxyRes.statusMessage}`, proxyRes.headers))
      proxyRes.pipe(res)
    }
  } else {
    res.statusCode = proxyRes.statusCode
    for (const [ key, value ] of Object.entries(headers)) {
      res.setHeader(key, value)
    }
    proxyRes.pipe(res)
  }
}

function onProxyReqUpgrade (proxyRes, proxySocket, proxyHead) {
  const res = this[kRes]

  res[kProxySocket] = proxySocket
  proxySocket[kRes] = res

  setupSocket(proxySocket)

  if (proxyHead && proxyHead.length) {
    proxySocket.unshift(proxyHead)
  }

  res.write(createHttpHeader('HTTP/1.1 101 Switching Protocols', proxyRes.headers))

  proxySocket
    .on('error', onProxyResError)
    .on('close', onProxyResAborted)
    .pipe(res)
    .pipe(proxySocket)
}

function onProxyResError (err) {
  err.statusCode = 502
  onComplete.call(this, err)
}

function onProxyResAborted () {
  onComplete.call(this, new HttpError('proxy aborted', 'ECONNRESET', 502))
}

function createHttpHeader (line, headers) {
  let head = line
  for (const [ key, value ] of Object.entries(headers)) {
    if (!Array.isArray(value)) {
      head += `\r\n${key}: ${value}`
    } else {
      for (let i = 0; i < value.length; i++) {
        head += `\r\n${key}: ${value[i]}`
      }
    }
  }
  head += '\r\n\r\n'
  return Buffer.from(head, 'ascii')
}

function getRequestHeaders (req, proxyName) {
  const headers = {}
  for (const [ key, value ] of Object.entries(req.headers)) {
    if (key.charAt(0) !== ':' && key !== 'host') {
      headers[key] = value
    }
  }

  // TODO(fix): <host> [ ":" <port> ] vs <pseudonym>
  // See, https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Via.
  if (proxyName) {
    if (headers[VIA]) {
      for (const name of headers[VIA].split(',')) {
        if (name.endsWith(proxyName)) {
          throw new HttpError('loop detected', null, 508)
        }
      }
      headers[VIA] += `,`
    } else {
      headers[VIA] = ''
    }

    headers[VIA] += `${req.httpVersion} ${req.proxyName}`
  }

  function printIp (address, port) {
    const isIPv6 = net.isIPv6(address)
    let str = `${address}`
    if (isIPv6) {
      str = `[${str}]`
    }
    if (port) {
      str = `${str}:${port}`
    }
    if (isIPv6) {
      str = `"${str}"`
    }
    return str
  }

  const forwarded = [
    `by=${printIp(req.socket.localAddress, req.socket.localPort)}`,
    `for=${printIp(req.socket.remoteAddress, req.socket.remotePort)}`,
    `proto=${req.socket.encrypted ? 'https' : 'http'}`,
    `host=${printIp(req.headers[AUTHORITY] || req.headers[HOST] || '')}`
  ].join('; ')

  if (headers[FORWARDED]) {
    headers[FORWARDED] += `, ${forwarded}`
  } else {
    headers[FORWARDED] = `${forwarded}`
  }

  return setupHeaders(headers)
}

function setupSocket (socket) {
  socket.setTimeout(0)
  socket.setNoDelay(true)
  socket.setKeepAlive(true, 0)
}

function setupHeaders (headers) {
  const connection = headers[CONNECTION]

  if (connection && connection !== CONNECTION && connection !== KEEP_ALIVE) {
    for (const name of connection.split(',')) {
      delete headers[name.trim()]
    }
  }

  // Remove hop by hop headers
  delete headers[CONNECTION]
  delete headers[PROXY_CONNECTION]
  delete headers[KEEP_ALIVE]
  delete headers[PROXY_AUTHENTICATE]
  delete headers[PROXY_AUTHORIZATION]
  delete headers[TE]
  delete headers[TRAILER]
  delete headers[TRANSFER_ENCODING]
  delete headers[UPGRADE]

  delete headers[HTTP2_SETTINGS]

  return headers
}

class HttpError extends Error {
  constructor (msg, code, statusCode) {
    super(msg)
    this.code = code
    this.statusCode = statusCode || 500
  }
}
